### PR TITLE
Possible implementation of Monadic Lazy module

### PR DESCRIPTION
**Pull request by austindd**
I just wanted to create this PR in case this was a desirable module. I'm still not sure if lazy evaluation makes sense as a monad, since the `join` operation essentially forces the evaluation of a lazy computation. But perhaps that is precisely what `join` and `bind` *should* mean in this context. I would appreciate any feedback on this.

The main reason I question the semantics is that most of the standard type-class functions (`map`, `bind`, `pure`, `apply`, etc.) can have alternate implementations where the computation is evaluated eagerly and then lifted back into the monadic context:
```ocaml
let map: ('a => 'b, t('a)) => t('b) = (aToB, mA) => pure(aToB(join(mA)));
```
I think this breaks some fundamental assumptions about the meaning of these functions, which is why I refrain from implementing the normal functions in that way, and instead provide alternative functions (with the same type signature), which DO force eager evaluation.

That's fine, we can have lazily evaluated monadic operations, alongside their eager counterparts. But it begs the question: is this really a monad?

I think the problem is that the `join` operation logically makes sense for both `t(t('a)) => t('a)` as well as `t('a) => 'a`. This is what allows the implicit `pure << join` operation to work inside the other functions, and this is not necessarily achievable with other monads. I would think taking advantage of the latter type to eagerly evaluate computations inside of `apply`, `map`, etc. would actually break the monad laws, right?

If so, then the answer might just be: "don't do that." But Idk...

Here is a relevant discussion on StackOverflow on this topic:
https://stackoverflow.com/questions/44289542/can-lazy-evaluation-be-implemented-by-a-monadic-type

### CODE REVIEW

**Comment by austindd** [comment]
Also, I forgot to mention two other concerns:

1. This module could also be implemented by wrapping the built-in `Lazy` module. This would allow it to interop with existing code that uses that module, and also allow use of the built-in `lazy` keyword. But the built-in functions throw exceptions and handles errors in a way that might not be ideal. I implemented the basic idea in a separate file named `Relude_Lazy2.re`.

2. Speaking of error handling, I suppose this could also be implemented as a bifunctor, if lazily computed values were intended to be a `Result.t('a, 'e)`. If we wanted it to wrap the built-in Lazy module, this would also give us a way to catch and return the possible exceptions into a `Result.t`.

**Comment by andywhite37** [comment]
Hi @austindd - Thanks for the PR and for starting this conversation. Yes, I think this would be a valuable module, and we should get this in, but I had a little feedback. I was looking at some of the prior art in purescript [Data.Lazy](https://github.com/purescript/purescript-lazy/blob/v3.0.0/src/Data/Lazy.purs) and other things like Scala cats [Eval](https://typelevel.org/cats/datatypes/eval.html).

One aspect that I think is worth adding here is memoization of the result.  The purescript lib seems to just memoize by default, and the cats `Eval` offers the possibility of memoization and also the `now`/`later`/`always` semantics. The purescript lib uses a JavaScript FFI `defer` to handle the memoization. This [foreign import defer](https://github.com/purescript/purescript-lazy/blob/v3.0.0/src/Data/Lazy.purs#L26) is just a binding to the [JS code here](https://github.com/purescript/purescript-lazy/blob/v3.0.0/src/Data/Lazy.js#L3-L25). The `force` function also appears to be a FFI to JS code. In Reason, I think we can just use `ref`s to handle the memoization, rather than JS bindings.

I think it would be cool to try to replicate the cats `Eval` with `now`/`later`/`always` semantics - that would make the type more flexible.

Cats `Eval` talks about implementing `bind` via trampolining for stack safety, but the purescript version seems to be simpler, and is quite easy to understand.  I think the simpler purescript version would be fine to start with I think - we can explore trampolining later.

As for the things like `map`/`apply`/`bind`/etc. I think it should be possible to implement those without necessarily forcing the computation at each step.  E.g. `map` could probably be implemented via function composition keeping the evaluation lazy. `flatMap`/`bind` will require a `force`, but that can be done inside a new thunk.

I don't think we really need/want the versions of those functions that `force` the evaluation each time, like `forceMap`, etc. I think the pattern would typically be to build up some lazy computation, then force it at the end.

In terms of the OCaml `Lazy`, we could probably handle that using a function in `Relude_Lazy` using a function like `fromOCamlLazy` or something?

The other thing that purescript has is a `Lazy` typeclass, which is used to abstract lazy things. I don't think we necessarily need to add this now - I think the lazy data type would be a good start.

To summarize, maybe we can look into these things:

- [ ] Explore using `now`/`later`/`always` semantics for handling the lazy evaluation
- [ ] Support memoization of the result in cases where it makes sense (e.g. in `now` and `later`, but not for `always`)?
- [ ] Look into implementing `map`, `apply`, `bind` that don't force the evaluation (e.g. via function composition)
- [ ] I think we can probably remove the `forceMap`/`forceFlatMap` etc. functions, and just have the `force` function to use to force a larger composed lazy expression.
- [ ] we could try looking into a trampolining implementation of `bind` (optional)
- [ ] See if we can support OCaml `Lazy` using just a simple conversion function like `fromOCamlLazy` and `toOCamlLazy`?

**Comment by austindd** [comment]
I agree with those points. I really like the Cats Eval semantics. I think we could achieve that pretty easily with a little time and adequate testing.

I looked into the stack-safety issue. I think I managed to get a stack-safe implementation of `bind`, but I'm not sure if I'm doing it the best way. I'll push another commit to demonstrate soon.

I wonder if it's possible to implement a stack-safe implementation of `map`. It seems impossible to me, if we want the computation to be deferred. `bind` has the nice property of forcing the computation at the `join`/`flatten` step, but `map` doesn't have that luxury.

Maintaining laziness with `map` probably doesn't work unless you have some kind of sentinel to force computations *later*...

I'm thinking we could implement a linked list of thunks which each decrement the countdown variable of the next one in the chain, and incrementally force their computations when their respective counters reach 0. This would prevent the undesirable "all-at-once" computation of all the thunks, but would still provide stack-safety, with some moderate performance cost at high stack sizes.

**Comment by andywhite37** [comment]
I'd advise just checking out how purescript Lazy is implemented as a starting point.

Based on the purescript implementation, this is what I would envision for `Lazy`. I don't believe `map` and `flatMap` are stack safe below, but I feel that this is a good first step, and we can explore other ideas later/as needed.

[Try reason](https://reasonml.github.io/try?rrjsx=true&reason=DYUwLgBAFAfDCUEC80BmAaCBzTBDRSM2UqU+8A3AFCiRQA89BamOE+yRpWZ8lVVALYB7ACYBXUBAAyuAF4BPZBADeVCBDAKADiE1QA5BxQAfGfIVRxAOwCWkQhCP91EWhFEhUIAE4AuaBt7TidjIjBDY2xwAEEQtQ0Nd0EQETiUHy8oADlhaxB+RLdwaLAYgFlU4Vs5EFFlKAIiAGcAd3sAYwALKBS0gD1EBKKIMwBlYRSoADdcYHECkNn5kFcis1z8+LWR91xlLFjG6hHEvuE4vxQJqfwT04gAKWaAOmBhLAAmKAAiWUUIMsFhAOpNtOIwHV2NZ6ucanU-D88IUHrh7kUAL47LFFf6WQ5lSoieGiFFY1zuVDCHwdEABCLOEJGZSoGyuMx4khNCCkFEUkrgzLKfaOTzeHxQRohO6ufmQQS4bQBZkGABGL2gzMcaswDPIIQZquYZAAKsIAEKYYAWGLc4YaMW+SXc62KOImLjU2mjIi4M3msn3OU861gcqK5X7NUayIGwxG3WRPhxtXGv3CPGWtw2u07R0SqWOdOZqCuhTuz00xYenlegr3LEY6g0EplsZgHy2axYZR4l7550hH4AFk+Pxc7jLAElrA5s4p253uz7zIoXgrtFAu2AAPrCVA75odrtYFwAegAVBATV1bM1a1X72Auno28fl7gYZoX-OFDPIBeZ5UM8bwfL8ABiXonhA24QAAtL+i7QTWfYblus57geR5Lqe47UJSdaiP+yjTrOK59lSVbUCB7w8NhJ6YTu24kIR-58M2l4QLkkL0re96iMIIDNNYBiQJkcGgoI4KQt+ej0d2mCqmeHQwZArS4PecyZLgohKJRtL1J+sJVCSVC4KqwjTHogHAa8tEQVBy7yVgLyuXhLaQPpdRIcuKBvjh5EWC8XnUXZYFeaIPmnhxV7cXS353h4gnCaJEDiZJ0l6M+r4WBAzmKRCslpUJkhgNBZYwWRiUZRCdRUKqIAdLg4jNFlfEwZpED5K0v4gnkkIAB5qfYXRFRuPI2B0ZV5BANk0WBPyQTS0HORA7TPhA424FguBdu5BFVkRs6fCRFhRYFa5odujHOYgKFBSFtmgTwEX-p8lBAA)

```reason
module Lazy = {
  type t('a) = | Lazy(unit => 'a);

  let defer: (unit => 'a) => t('a) = getA => {
    let memoA = ref(None);
    let getAMemoized = () => switch(memoA^) {
      | Some(value) => value
      | None => {
        let a = getA();
        memoA := Some(a);
        Js.log2("Lazy value computed and memoized:", a);
        a;
      }
    }
    Lazy(getAMemoized);
  }

  let force: t('a) => 'a = fun
  | Lazy(f) => f();
  
  let pure = a => defer(() => a);
  
  let map: 'a 'b. ('a => 'b, t('a)) => t('b) = (aToB, lazyA) => {
    defer(() => lazyA |> force |> aToB);
  };
  
  let flatMap: 'a 'b. ('a => t('b), t('a)) => t('b) = (aToLazyB, lazyA) => {
    defer(() => aToLazyB(lazyA |> force) |> force);
  }
};

let lazyString = Lazy.defer(() => "42");

let lazyInt = lazyString |> Lazy.map(int_of_string);

/* This forces the lazyString and the lazyInt */
Js.log("Forcing int - lazyString |> Lazy.map(int_of_string)");
let forcedInt = lazyInt |> Lazy.force;
Js.log(string_of_int(forcedInt));

/* Note: this doesn't re-compute the string, b/c it was already forced and memoized
above */
Js.log("Forcing string...");
let forcedString = lazyString |> Lazy.force;
Js.log(forcedString);

/* Note: this doesn't re-compute the lazy string, but the resulting lazy int is computed
because this is a new lazy context with the map function */
Js.log("Forcing string with map again");
let forcedInt2 = lazyString |> Lazy.map(int_of_string) |> Lazy.force;
Js.log(forcedInt2);
```

As far as stack safety for `map`, I'm not sure. I think in practice stack safety usually comes up more with `flatMap`/`bind`, because of the nature of monads - like you build up chains of evaluations, but each step must be evaluated before the following step, so you tend to get these chains of computations, whereas with map, you don't tend to see chains of thousands of map calls, and you can often compose map functions inline b/c they don't need the forced evaluation like monads do. That said, nobody is stopping you from chaining tons of map calls, and I don't know what the current thinking is on stack safety for map. There's some thinking going on in FP land about stack safety/trampolining/etc. with function composition:  https://medium.com/@safareli/stack-safe-function-composition-85d61feee37e

I would suggest digging into some of this stuff in purescript/Scala/etc. before trying to invent new ideas. The thing you mentioned about a linked list of thunks is quite similar to the concept of Free Monads - where you can capture monadic computations in a data structure, and interpret them at a later time - that would be worth reading about if you haven't dug into those yet. This is a good blog about that:  https://www.parsonsmatt.org/2017/09/22/what_does_free_buy_us.html

**Comment by austindd** [comment]
@andywhite37 I totally agree. I think maybe we should close this PR for now until we have a better plan for the implementation. Thanks for the reading material as well. That's very helpful!

**Comment by andywhite37** [comment]
Ok, I'll close for now, but we can re-open later if we want to revisit.
